### ISSUE TITLE
1.2.840.113549.1.1.4 Signature not available for BCJSSE in fips mode

### ISSUE THREAD

**Issue by jkakavas** [elastic/elasticsearch#66821]
**Build scan**:
https://gradle-enterprise.elastic.co/s/sl7mn5m3gqtts/console-log?task=:client:sniffer:test

**Repro line**:

**Reproduces locally?**:

**Applicable branches**:

**Failure history**:

**Failure excerpt**:

```
: Test class org.elasticsearch.client.sniff.SniffOnFailureListenerTests |  
-- | --
  | 2> Dec 26, 2020 5:01:56 AM org.bouncycastle.jsse.provider.ProvKeyManagerFactorySpi getDefaultKeyStore |  
  | 2> INFO: Initializing empty key store |  
  | 2> Dec 26, 2020 5:01:56 AM org.bouncycastle.jsse.provider.ProvTrustManagerFactorySpi getDefaultTrustStore |  
  | 2> INFO: Initializing with trust store at path: /dev/shm/elastic+elasticsearch+master+matrix-java-periodic-fips/ES_RUNTIME_JAVA/java11/nodes/general-purpose/client/sniffer/build/fips-resources/cacerts.bcfks |  
  | 2> java.lang.ExceptionInInitializerError |  
  | at __randomizedtesting.SeedInfo.seed([E9613EC44B53D6F:A6FD937FBFBB3E3D]:0) |  
  | at java.base/javax.crypto.ProviderVerifier.verify(ProviderVerifier.java:128) |  
  | at java.base/javax.crypto.JceSecurity.verifyProvider(JceSecurity.java:191) |  
  | at java.base/javax.crypto.JceSecurity.getVerificationResult(JceSecurity.java:217) |  
  | at java.base/javax.crypto.JceSecurity.getInstance(JceSecurity.java:141) |  
  | at java.base/javax.crypto.Mac.getInstance(Mac.java:273) |  
  | at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.calculateMac(Unknown Source) |  
  | at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.verifyMac(Unknown Source) |  
  | at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.engineLoad(Unknown Source) |  
  | at java.base/java.security.KeyStore.load(KeyStore.java:1479) |  
  | at org.bouncycastle.jsse.provider.ProvTrustManagerFactorySpi.getDefaultTrustStore(ProvTrustManagerFactorySpi.java:98) |  
  | at org.bouncycastle.jsse.provider.ProvSSLContextSpi.getDefaultTrustManagers(ProvSSLContextSpi.java:302) |  
  | at org.bouncycastle.jsse.provider.DefaultSSLContextSpi$LazyManagers.<clinit>(DefaultSSLContextSpi.java:57) |  
  | at org.bouncycastle.jsse.provider.DefaultSSLContextSpi.<init>(DefaultSSLContextSpi.java:86) |  
  | at org.bouncycastle.jsse.provider.BouncyCastleJsseProvider$7.createInstance(BouncyCastleJsseProvider.java:217) |  
  | at org.bouncycastle.jsse.provider.BouncyCastleJsseProvider$BcJsseService.newInstance(BouncyCastleJsseProvider.java:373) |  
  | at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:236) |  
  | at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:164) |  
  | at java.base/javax.net.ssl.SSLContext.getInstance(SSLContext.java:168) |  
  | at java.base/javax.net.ssl.SSLContext.getDefault(SSLContext.java:99) |  
  | at org.elasticsearch.client.RestClientBuilder.createHttpClient(RestClientBuilder.java:222) |  
  | at java.base/java.security.AccessController.doPrivileged(Native Method) |  
  | at org.elasticsearch.client.RestClientBuilder.build(RestClientBuilder.java:201) |  
  | at org.elasticsearch.client.sniff.SniffOnFailureListenerTests.testSetSniffer(SniffOnFailureListenerTests.java:49) |  
  |   |  
  | Caused by: |  
  | java.lang.SecurityException: Framework jar verification can not be initialized |  
  | at java.base/javax.crypto.JarVerifier.<clinit>(JarVerifier.java:198) |  
  | ... 23 more |  
  |   |  
  | Caused by: |  
  | java.security.PrivilegedActionException: java.security.NoSuchAlgorithmException: 1.2.840.113549.1.1.4 Signature not available |  
  | at java.base/java.security.AccessController.doPrivileged(Native Method) |  
  | at java.base/javax.crypto.JarVerifier.<clinit>(JarVerifier.java:166) |  
  | ... 23 more |  
  |   |  
  | Caused by: |  
  | java.security.NoSuchAlgorithmException: 1.2.840.113549.1.1.4 Signature not available |  
  | at java.base/java.security.Signature.getInstance(Signature.java:267) |  
  | at org.bouncycastle.jcajce.provider.X509CertificateObject.getSignatureFromProvider(Unknown Source) |  
  | at org.bouncycastle.jcajce.provider.X509CertificateObject.verify(Unknown Source) |  
  | at java.base/javax.crypto.JarVerifier.testSignatures(JarVerifier.java:735) |  
  | at java.base/javax.crypto.JarVerifier$1.run(JarVerifier.java:192) |  
  | at java.base/javax.crypto.JarVerifier$1.run(JarVerifier.java:167) |  
  | ... 25 more

```

It's not obvious what triggers this and it is only occured in certain tests and not always or when starting elasticsearch in fips mode. I presume it has to do with a combination of jdk provider and version.  @BigPandaToo has been seeing this locally too IIRC

It looks like we are hitting https://bugs.openjdk.java.net/browse/JDK-8198743 and the appropriate workaround would be to add the SunRsaSign provider too in the list of providers we pack for CI in fips_java.security file, but we should look into why this is not always triggered and the verification of what jar causes this ( I.e. where is the md5withrsa signature used)

**Comment by mayya-sharipova** [comment]
Recent failure today: https://gradle-enterprise.elastic.co/s/czg7xzhae3ek2

**Comment by ywangd** [comment]
This failure can be reproduced locally with an Oracle JVM using:
```
./gradlew :client:sniffer:test  -Dtests.fips.enabled=true -Druntime.java=11
```

The same "signature not available" failure can happen for many other tests when Oracle JVM is in use, e.g.:
```
./gradlew ':modules:transport-netty4:javaRestTest' --tests "org.elasticsearch.rest.Netty4HeadBodyIsEmptyIT.testTemplateExists"  -Druntime.java=15 -Dtests.fips.enabled=true
```

These tests fail to start because Oracle JVM uses hardcoded certificates with `md5WithRsa` signature, which is not allowed with BC Provider in `approved_only` mode. They do *not* fail with OpenJDK because the implementation of `javax.crypto.ProviderVerifier#verify` is different between Oracle JDK and OpenJDK. In short, OpenJDK is more lenient in JarFile verification which allows it to skip Security Provider verification if it does not need to write any permissions:
https://github.com/openjdk/jdk15u/blob/0934873e869c6f4b8acbd9508ab49141a66a9692/src/java.base/share/classes/javax/crypto/ProviderVerifier.java#L86-L91
The above code basically skips the signature verification step so that `javax.crypto.JarVerifier` is never used.

However, the same method has a different implementation in Oracle JDK (based on decompilation):
```java
void verify() throws IOException {
    if (!this.isTrustedModule()) {
        if (this.url.getProtocol().equalsIgnoreCase("jrt")) {
            throw new JarException("Non-Oracle JCE providers may not be linked into the image,they must be provided as signed JAR files.");
        } else {
            this.jarVerifier = new JarVerifier(this.url, this.savePerms);
            this.jarVerifier.verify();
        }
    }
}
```

The above code basically says if the Jar is not a trusted module, e.g. `bcfips`, the code goes ahead to verify the Jar signatures. But the `JarVerifier#testSignatures` method (invoked as part of `new JarVerifier(...)`) uses two hardcoded certificates with `md5WithRsa` signatures and this is where the code fails to proceed further. The relevant code are something like the follows (decompiled):
```java
private static void testSignatures(X509Certificate providerCert, CertificateFactory certFactory) throws Exception {
    ...
    // The following certs have hardcoded certificates with md5WithRsa signature
    X509Certificate[] certs = new X509Certificate[]{providerCert, parseCertificate(DSA, certFactory), parseCertificate("-----BEGIN CERTIFICATE-----\nMIIB4DCCAYoCAQEwDQYJKoZIhvcNAQEEBQAwezELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAkNBMRIwEAYDVQQHEwlDdXBlcnRpbm8xGTAXBgNVBAoTEFN1biBNaWNyb3N5c3RlbXMxFjAUBgNVBAsTDUphdmEgU29mdHdhcmUxGDAWBgNVBAMTD0pDRSBEZXZlbG9wbWVudDAeFw0wMjEwMzExNTI3NDRaFw0wNzEwMzExNTI3NDRaMHsxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTESMBAGA1UEBxMJQ3VwZXJ0aW5vMRkwFwYDVQQKExBTdW4gTWljcm9zeXN0ZW1zMRYwFAYDVQQLEw1KYXZhIFNvZnR3YXJlMRgwFgYDVQQDEw9KQ0UgRGV2ZWxvcG1lbnQwXDANBgkqhkiG9w0BAQEFAANLADBIAkEAo/4CddEOa3M6v9JFAhnBYgTq54Y30++F8yzCK9EeYaG3AzvzZqNshDy579647p0cOM/4VO6rU2PgbzgKXPcs8wIDAQABMA0GCSqGSIb3DQEBBAUAA0EACqPlFmVdKdYSCTNltXKQnBqss9GNjbnB+CitvWrwN+oOK8qQpvV+5LB6LruvRy6zCedCV95Z2kXKg/Fnj0gvsg==\n-----END CERTIFICATE-----", certFactory), parseCertificate("-----BEGIN CERTIFICATE-----\nMIIB4DCCAYoCAQIwDQYJKoZIhvcNAQEEBQAwezELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAkNBMRIwEAYDVQQHEwlDdXBlcnRpbm8xGTAXBgNVBAoTEFN1biBNaWNyb3N5c3RlbXMxFjAUBgNVBAsTDUphdmEgU29mdHdhcmUxGDAWBgNVBAMTD0pDRSBEZXZlbG9wbWVudDAeFw0wMjEwMzExNTI3NDRaFw0wNzEwMzExNTI3NDRaMHsxCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTESMBAGA1UEBxMJQ3VwZXJ0aW5vMRkwFwYDVQQKExBTdW4gTWljcm9zeXN0ZW1zMRYwFAYDVQQLEw1KYXZhIFNvZnR3YXJlMRgwFgYDVQQDEw9KQ0UgRGV2ZWxvcG1lbnQwXDANBgkqhkiG9w0BAQEFAANLADBIAkEAr1OSXaOzpnVoqL2LqS5+HLy1kVvBwiM/E5iYT9eZaghE8qvF+4fETipWUNTWCQzHR4cDJGJOl9Nm77tELhES4QIDAQABMA0GCSqGSIb3DQEBBAUAA0EAL+WcVFyj+iXlEVNVQbNOOUlWmlmXGiNKKXnIdNcc1ZUyi+JW0zmlfZ7iU/eRYhEEJBwdrUoyiGOGLo7pi6JzAA==\n-----END CERTIFICATE-----", certFactory)};
    ...
    publicKeys[2] = certs[2].getPublicKey();
    publicKeys[3] = publicKeys[2];
 
    for(int i = 0; i < 12; ++i) {
        try {
            certs[k].verify(publicKeys[k]);  // <---- ERROR HERE
        }
        ...
    }
}
```

This also means Oracle JVM essentially cannot run in FIPS `approved_only` mode with `bcfips` as the only provider in Production! The workaround is to add `SunRsaSign` into the security file. But strictly speaking, I suspect the whole setup is then no longer really FIPS compliant since it allows md5. Our default distribution ships with `AdoptOpenJDK`, so it does not have this problem. This is also why I was able to run elasticsearch in FIPS mode locally because it was using the bundled JDK. But it fails as soon as I explicity specify the `JAVA_HOME` pointing to an Oracle version.

**Comment by jkakavas** [comment]
This is still failing, see https://gradle-enterprise.elastic.co/s/pnfomh653lls2 , reopening to investigate

**Comment by jkakavas** [comment]
This broke again because of #65596. `JavaInstallation#getImplementation` was changed to `InstallationLocation#getDisplayName` for the value of `BuildParams.runtimeJavaDetails` and the latter doesn't include the vendor name. I've pinged es-delivery folks to see if there is something else we could use to determine if the jdk used is an oracle one or not

**Comment by jkakavas** [comment]
Resolved (again) in https://github.com/elastic/elasticsearch/pull/67417

**Comment by astefan** [comment]
Spotted another failure this morning. Build scan https://gradle-enterprise.elastic.co/s/s26ii6mphzji6

```
org.elasticsearch.client.RestClientBuilderTests > testBuildCloudIdWithPort FAILED
    java.lang.ExceptionInInitializerError
        at __randomizedtesting.SeedInfo.seed([F3355D2B4A1429D:1D9208471D8BE78C]:0)
        at java.base/javax.crypto.ProviderVerifier.verify(ProviderVerifier.java:128)
        at java.base/javax.crypto.JceSecurity.verifyProvider(JceSecurity.java:191)
        at java.base/javax.crypto.JceSecurity.getVerificationResult(JceSecurity.java:217)
        at java.base/javax.crypto.JceSecurity.getInstance(JceSecurity.java:141)
        at java.base/javax.crypto.Mac.getInstance(Mac.java:273)
        at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.calculateMac(Unknown Source)
        at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.verifyMac(Unknown Source)
        at org.bouncycastle.jcajce.provider.ProvBCFKS$BCFIPSKeyStoreSpi.engineLoad(Unknown Source)
        at java.base/java.security.KeyStore.load(KeyStore.java:1479)
        at org.bouncycastle.jsse.provider.ProvTrustManagerFactorySpi.getDefaultTrustStore(ProvTrustManagerFactorySpi.java:98)
        at org.bouncycastle.jsse.provider.ProvSSLContextSpi.getDefaultTrustManagers(ProvSSLContextSpi.java:302)
        at org.bouncycastle.jsse.provider.DefaultSSLContextSpi$LazyManagers.<clinit>(DefaultSSLContextSpi.java:57)
        at org.bouncycastle.jsse.provider.DefaultSSLContextSpi.<init>(DefaultSSLContextSpi.java:86)
        at org.bouncycastle.jsse.provider.BouncyCastleJsseProvider$7.createInstance(BouncyCastleJsseProvider.java:217)
        at org.bouncycastle.jsse.provider.BouncyCastleJsseProvider$BcJsseService.newInstance(BouncyCastleJsseProvider.java:373)
        at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
        at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
        at java.base/javax.net.ssl.SSLContext.getInstance(SSLContext.java:168)
        at java.base/javax.net.ssl.SSLContext.getDefault(SSLContext.java:99)
        at org.elasticsearch.client.RestClientBuilder.createHttpClient(RestClientBuilder.java:222)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at org.elasticsearch.client.RestClientBuilder.build(RestClientBuilder.java:201)
        at org.elasticsearch.client.RestClientBuilderTests.testBuildCloudIdWithPort(RestClientBuilderTests.java:204)

        Caused by:
        java.lang.SecurityException: Framework jar verification can not be initialized
            at java.base/javax.crypto.JarVerifier.<clinit>(JarVerifier.java:198)
            ... 23 more

            Caused by:
            java.security.PrivilegedActionException: java.security.NoSuchAlgorithmException: 1.2.840.113549.1.1.4 Signature not available
                at java.base/java.security.AccessController.doPrivileged(Native Method)
                at java.base/javax.crypto.JarVerifier.<clinit>(JarVerifier.java:166)
                ... 23 more

                Caused by:
                java.security.NoSuchAlgorithmException: 1.2.840.113549.1.1.4 Signature not available
                    at java.base/java.security.Signature.getInstance(Signature.java:267)
                    at org.bouncycastle.jcajce.provider.X509CertificateObject.getSignatureFromProvider(Unknown Source)
                    at org.bouncycastle.jcajce.provider.X509CertificateObject.verify(Unknown Source)
                    at java.base/javax.crypto.JarVerifier.testSignatures(JarVerifier.java:735)
                    at java.base/javax.crypto.JarVerifier$1.run(JarVerifier.java:192)
                    at java.base/javax.crypto.JarVerifier$1.run(JarVerifier.java:167)
                    ... 25 more
```

**Comment by ywangd** [comment]
This happens again on CI for oracle jdk 1.8.0_301:
* https://gradle-enterprise.elastic.co/s/vqdx6hotfgt4y
* https://gradle-enterprise.elastic.co/s/zkb66s2qs2fe6

Also [reproducible](https://gradle-enterprise.elastic.co/s/iwuwcqfwjokt6) locally with oracle jdk 1.8.0_241 with
> ./gradlew ':x-pack:plugin:ccr:test' --tests "org.elasticsearch.xpack.monitoring.collector.ccr.StatsCollectorTests.testShouldCollectReturnsFalseIfCCRIsNotAllowed" -Dtests.seed=2C4A2EE151085BB -Dtests.locale=ar-DZ -Dtests.timezone=Atlantic/St_Helena -Druntime.java=8 -Dtests.fips.enabled=true

**Comment by ywangd** [comment]
Should be fixed by #77123
